File: sdks/java/flux-sdk/src/main/java/io/flux/sdk/RawReader.java
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (c) 2025 Nikhil Simha Raprolu

package io.flux.sdk;

import io.flux.sdk.proto.AuthRequest;
import io.flux.sdk.proto.ClientMessage;
import io.flux.sdk.proto.RawReadRequest;
import io.flux.sdk.proto.RawReadResponse;
import io.flux.sdk.proto.RawSegment;
import io.flux.sdk.proto.ServerMessage;
import org.java_websocket.client.WebSocketClient;
import org.java_websocket.handshake.ServerHandshake;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.URI;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;

/**
 * Zero-copy reader client.
 * <p>
 * Fetches compressed FL segments from the broker as-is and decodes them
 * locally via {@link SegmentDecoder}. Unlike {@link GroupReader}, raw reads
 * address offsets directly and do not participate in reader groups.
 */
public class RawReader implements AutoCloseable {
    private static final Logger log = LoggerFactory.getLogger(RawReader.class);
    private static final int RESPONSE_QUEUE_CAPACITY = 1000;

    private final ReaderConfig config;
    private final FluxWebSocketClient client;
    private final BlockingQueue<byte[]> responseQueue;

    /** Offset range covered by a single decoded segment. */
    public static class SegmentRange {
        private final int topicId;
        private final int schemaId;
        private final long startOffset;
        private final long endOffset;

        public SegmentRange(int topicId, int schemaId, long startOffset, long endOffset) {
            this.topicId = topicId;
            this.schemaId = schemaId;
            this.startOffset = startOffset;
            this.endOffset = endOffset;
        }

        public int getTopicId() { return topicId; }
        public int getSchemaId() { return schemaId; }
        public long getStartOffset() { return startOffset; }
        public long getEndOffset() { return endOffset; }
    }

    /** Result of a single raw read: decoded records, segment ranges, and the high watermark. */
    public static class RawReadBatch {
        private final List<SegmentDecoder.DecodedRecord> records;
        private final List<SegmentRange> segments;
        private final long highWatermark;

        public RawReadBatch(List<SegmentDecoder.DecodedRecord> records,
                            List<SegmentRange> segments,
                            long highWatermark) {
            this.records = records;
            this.segments = segments;
            this.highWatermark = highWatermark;
        }

        public List<SegmentDecoder.DecodedRecord> getRecords() { return records; }
        public List<SegmentRange> getSegments() { return segments; }
        public long getHighWatermark() { return highWatermark; }
    }

    private RawReader(ReaderConfig config, FluxWebSocketClient client, BlockingQueue<byte[]> responseQueue) {
        this.config = config;
        this.client = client;
        this.responseQueue = responseQueue;
    }

    public static RawReader connect(ReaderConfig config) throws FluxException {
        try {
            URI uri = new URI(config.getUrl());
            BlockingQueue<byte[]> responseQueue = new LinkedBlockingQueue<>(RESPONSE_QUEUE_CAPACITY);
            FluxWebSocketClient client = new FluxWebSocketClient(uri, responseQueue);

            if (!client.connectBlocking(config.getTimeout().toMillis(), TimeUnit.MILLISECONDS)) {
                throw new FluxException.ConnectionException("Failed to connect to " + config.getUrl());
            }

            RawReader reader = new RawReader(config, client, responseQueue);
            if (config.getApiKey() != null) {
                reader.authenticate(config.getApiKey());
            }
            return reader;
        } catch (Exception e) {
            if (e instanceof FluxException) {
                throw (FluxException) e;
            }
            throw new FluxException.ConnectionException("Failed to connect", e);
        }
    }

    private void authenticate(String apiKey) throws FluxException {
        ClientMessage authMessage = ClientMessage.newBuilder()
                .setAuth(AuthRequest.newBuilder().setApiKey(apiKey).build())
                .build();
        client.send(authMessage.toByteArray());

        try {
            byte[] response = responseQueue.poll(config.getTimeout().toMillis(), TimeUnit.MILLISECONDS);
            if (response == null) {
                throw new FluxException.TimeoutException("Authentication timeout");
            }

            ServerMessage envelope = ServerMessage.parseFrom(response);
            if (envelope.getMessageCase() != ServerMessage.MessageCase.AUTH) {
                throw new FluxException.ProtocolException("Unexpected auth response");
            }

            io.flux.sdk.proto.AuthResponse auth = envelope.getAuth();
            if (!auth.getSuccess()) {
                throw new FluxException.AuthenticationException(
                        "Auth failed (" + auth.getErrorCode() + "): " + auth.getErrorMessage()
                );
            }
            log.debug("Authentication successful");
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new FluxException.ConnectionException("Interrupted during auth", e);
        } catch (com.google.protobuf.InvalidProtocolBufferException e) {
            throw new FluxException.ProtocolException("Failed to decode auth response", e);
        }
    }

    /**
     * Read compressed segments starting at {@code offset}, verify each segment's
     * CRC-32, and decode the contained records.
     *
     * @param topicId  topic to read from
     * @param offset   first offset to read
     * @param maxBytes budget in compressed bytes
     */
    public RawReadBatch rawRead(int topicId, long offset, int maxBytes) throws FluxException {
        RawReadRequest req = RawReadRequest.newBuilder()
                .setTopicId(topicId)
                .setOffset(offset)
                .setMaxBytes(maxBytes)
                .build();
        client.send(ClientMessage.newBuilder().setRawRead(req).build().toByteArray());

        try {
            byte[] response = responseQueue.poll(config.getTimeout().toMillis(), TimeUnit.MILLISECONDS);
            if (response == null) {
                throw new FluxException.TimeoutException("Raw read timeout");
            }

            ServerMessage envelope = ServerMessage.parseFrom(response);
            if (envelope.getMessageCase() != ServerMessage.MessageCase.RAW_READ) {
                throw new FluxException.ProtocolException("Unexpected response type or empty response");
            }

            RawReadResponse resp = envelope.getRawRead();
            if (!resp.getSuccess()) {
                throw new FluxException.ProtocolException(
                        "Raw read failed (" + resp.getErrorCode() + "): " + resp.getErrorMessage()
                );
            }

            List<SegmentDecoder.DecodedRecord> records = new ArrayList<>();
            List<SegmentRange> segments = new ArrayList<>(resp.getSegmentsCount());
            for (RawSegment segment : resp.getSegmentsList()) {
                records.addAll(SegmentDecoder.decode(
                        segment.getPayload().toByteArray(),
                        Integer.toUnsignedLong(segment.getCrc32())
                ));
                segments.add(new SegmentRange(
                        segment.getTopicId(),
                        segment.getSchemaId(),
                        segment.getStartOffset(),
                        segment.getEndOffset()
                ));
            }

            return new RawReadBatch(records, segments, resp.getHighWatermark());
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new FluxException.ConnectionException("Interrupted", e);
        } catch (com.google.protobuf.InvalidProtocolBufferException e) {
            throw new FluxException.ProtocolException("Failed to decode response", e);
        }
    }

    @Override
    public void close() {
        client.close();
    }

    private static class FluxWebSocketClient extends WebSocketClient {
        private static final Logger log = LoggerFactory.getLogger(FluxWebSocketClient.class);
        private final BlockingQueue<byte[]> responseQueue;

        public FluxWebSocketClient(URI serverUri, BlockingQueue<byte[]> responseQueue) {
            super(serverUri);
            this.responseQueue = responseQueue;
        }

        @Override
        public void onOpen(ServerHandshake handshake) {
            log.debug("WebSocket connected");
        }

        @Override
        public void onMessage(String message) {
            log.warn("Received text message, expected binary");
        }

        @Override
        public void onMessage(ByteBuffer bytes) {
            byte[] data = new byte[bytes.remaining()];
            bytes.get(data);
            responseQueue.offer(data);
        }

        @Override
        public void onClose(int code, String reason, boolean remote) {
            log.debug("WebSocket closed: {} - {}", code, reason);
        }

        @Override
        public void onError(Exception ex) {
            log.error("WebSocket error", ex);
        }
    }
}

File: sdks/python/flux/raw_reader.py
# SPDX-License-Identifier: AGPL-3.0-only
# Copyright (c) 2025 Nikhil Simha Raprolu

"""Zero-copy reader: fetches compressed FL segments and decodes them locally."""

import asyncio
import logging
from dataclasses import dataclass
from typing import Optional

import websockets
from websockets.asyncio.client import ClientConnection

from flux.proto import flux_wire_pb2 as pb

from .exceptions import (
    AuthenticationException,
    ConnectionException,
    ProtocolException,
    TimeoutException,
)
from .reader import ReaderConfig
from .segment import decode_segment

logger = logging.getLogger(__name__)


@dataclass
class RawSegmentBatch:
    """Records decoded from a single FL segment, with its offset range."""

    topic_id: int
    schema_id: int
    start_offset: int
    end_offset: int
    records: list[tuple[Optional[bytes], bytes]]


@dataclass
class RawReadResult:
    """Result of a raw read: decoded segments plus the topic high watermark."""

    segments: list[RawSegmentBatch]
    high_watermark: int

    @property
    def records(self) -> list[tuple[Optional[bytes], bytes]]:
        """All (key, value) pairs across segments, in offset order."""
        return [record for segment in self.segments for record in segment.records]

    @property
    def next_offset(self) -> Optional[int]:
        """Offset to request next, or None if this read returned no segments."""
        return self.segments[-1].end_offset if self.segments else None


class RawReader:
    """Reader that requests raw FL segments and decodes records client-side.

    Unlike GroupReader, reads are addressed by explicit offset and do not
    participate in a reader group. `max_bytes` budgets *compressed* bytes.
    """

    def __init__(self, ws: ClientConnection, config: ReaderConfig):
        self._ws = ws
        self._config = config

    @classmethod
    async def connect(cls, config: ReaderConfig) -> "RawReader":
        try:
            ws = await asyncio.wait_for(
                websockets.connect(config.url),
                timeout=config.timeout,
            )
        except asyncio.TimeoutError:
            raise TimeoutException("Connection timeout")
        except Exception as e:
            raise ConnectionException(f"Failed to connect: {e}")

        reader = cls(ws, config)
        if config.api_key:
            await reader._authenticate(config.api_key)
        return reader

    async def _authenticate(self, api_key: str) -> None:
        envelope = pb.ClientMessage()
        envelope.auth.api_key = api_key
        await self._ws.send(envelope.SerializeToString())

        try:
            response = await asyncio.wait_for(
                self._ws.recv(),
                timeout=self._config.timeout,
            )
        except asyncio.TimeoutError:
            raise TimeoutException("Authentication timeout")

        if not isinstance(response, bytes):
            raise ProtocolException("Expected binary response")

        server_msg = pb.ServerMessage()
        server_msg.ParseFromString(response)
        if server_msg.WhichOneof("message") != "auth":
            raise ProtocolException("Unexpected auth response")

        auth_resp = server_msg.auth
        if not auth_resp.success:
            raise AuthenticationException(auth_resp.error_message)

        logger.debug("Authentication successful")

    async def read(
        self,
        offset: int,
        max_bytes: Optional[int] = None,
    ) -> RawReadResult:
        """Read raw segments starting at `offset` and decode their records."""
        req = pb.RawReadRequest(
            topic_id=self._config.topic_id,
            offset=offset,
            max_bytes=max_bytes if max_bytes is not None else self._config.max_bytes,
        )
        envelope = pb.ClientMessage()
        envelope.raw_read.CopyFrom(req)
        await self._ws.send(envelope.SerializeToString())

        try:
            response = await asyncio.wait_for(
                self._ws.recv(),
                timeout=self._config.timeout,
            )
        except asyncio.TimeoutError:
            raise TimeoutException("Raw read timeout")

        if not isinstance(response, bytes):
            raise ProtocolException("Expected binary response")

        server_msg = pb.ServerMessage()
        server_msg.ParseFromString(response)
        if server_msg.WhichOneof("message") != "raw_read":
            raise ProtocolException("Unexpected response type or empty response")

        resp = server_msg.raw_read
        if not resp.success:
            raise ProtocolException(
                f"Raw read failed ({resp.error_code}): {resp.error_message}"
            )

        segments = [
            RawSegmentBatch(
                topic_id=seg.topic_id,
                schema_id=seg.schema_id,
                start_offset=seg.start_offset,
                end_offset=seg.end_offset,
                records=decode_segment(seg.payload, expected_crc=seg.crc32),
            )
            for seg in resp.segments
        ]
        return RawReadResult(segments=segments, high_watermark=resp.high_watermark)

    async def close(self) -> None:
        await self._ws.close()

    async def __aenter__(self) -> "RawReader":
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb) -> None:
        await self.close()
